pkg/engine/internal/planner/physical/catalog.go

```diff
@@ -145,110 +125,175 @@ func (c *MetastoreCatalog) resolveShardDescriptorsWithIndex(selector Expression,
+	req := CatalogRequest{
+		Kind:       CatalogRequestKindResolveShardDescriptorsWithShard,
+		Selector:   cloneExpressions([]Expression{selector})[0],
+		Predicates: cloneExpressions(predicates),
+		Shard:      shard,
+		From:       from,
+		Through:    through,
 	}
-	if l.ValueType() != types.Loki.String {
-		return "", fmt.Errorf("literal type is not a string, got %v", l.ValueType())
+
+	key := newCatalogRequestKey(req)
+	if _, ok := c.requests[key]; !ok {
+		c.requests[key] = req
 	}
-	return l.Value().(string), nil
+	return nil, nil
 }
 
-func convertColumnRef(expr Expression, allowAmbiguousColumnRefs bool) (string, error) {
-	ref, ok := expr.(*ColumnExpr)
-	if !ok {
-		return "", fmt.Errorf("expected column expression, got %T", expr)
+func (c UnresolvedCatalog) RequestsCount() int {
+	return len(c.requests)
+}
+
+func (c UnresolvedCatalog) Resolve(resolve func(CatalogRequest) (CatalogResponse, error)) (ResolvedCatalog, error) {
+	resolved := ResolvedCatalog{
+		responses: make(map[catalogRequestKey]CatalogResponse),
 	}
-	if !allowAmbiguousColumnRefs && ref.Ref.Type != types.ColumnTypeLabel {
-		return "", fmt.Errorf("column type is not a label, got %v", ref.Ref.Type)
+	for key, req := range c.requests {
+		resp, err := resolve(req)
+		if err != nil {
+			return resolved, fmt.Errorf("failed to resolve catalog response: %w", err)
+		}
+		resolved.responses[key] = resp
 	}
-	return ref.Ref.Column, nil
+
+	return resolved, nil
+}
+
+type ResolvedCatalog struct {
+	responses map[catalogRequestKey]CatalogResponse
 }
 
-func convertBinaryOp(t types.BinaryOp) (labels.MatchType, error) {
-	ty, ok := binOpToMatchTypeMapping[t]
+func (c ResolvedCatalog) ResolveShardDescriptorsWithShard(selector Expression, predicates []Expression, shard ShardInfo, from, through time.Time) ([]FilteredShardDescriptor, error) {
+	req := CatalogRequest{
+		Kind:       CatalogRequestKindResolveShardDescriptorsWithShard,
+		Selector:   selector,
+		Predicates: predicates,
+		Shard:      shard,
+		From:       from,
+		Through:    through,
+	}
+	reqKey := newCatalogRequestKey(req)
+	resp, ok := c.responses[reqKey]
 	if !ok {
-		return -1, fmt.Errorf("invalid binary operator for matcher: %v", t)
+		return nil, fmt.Errorf("catalog response missing for request %+v", reqKey)
+	}
+
+	switch resp.Kind {
+	case CatalogRequestKindResolveShardDescriptorsWithShard:
+		return resp.Descriptors, nil
+	default:
+		return nil, fmt.Errorf("unsupported response kind %d", resp.Kind)
+	}
+}
+
+// FilterDescriptorsForShard filters the section descriptors for a given shard.
+// It returns the locations, streams, and sections for the shard.
+// TODO: Improve filtering: this method could be improved because it doesn't resolve the stream IDs to sections, even though this information is available. Instead, it resolves streamIDs to the whole object.
+func FilterDescriptorsForShard(shard ShardInfo, sectionDescriptors []*metastore.DataobjSectionDescriptor) ([]FilteredShardDescriptor, error) {
```

Comment:
I don't think we use the ShardInfo anymore with the new scheduler. It might be time to remove it. We can do it after this is merged though.